rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,468 @@
+this may raise the priority of other existing API gaps for XLA.  For example, 
+the XLA version of `TensorStridedSliceUpdate` op does not currently support
+non-unit strides nor broadcasting.  It should otherwise work with Cloud TPUs.
+
+### Best Practices
+
+Due to copying of tensor values, the best practice will be to minimize the
+number of calls to the proposed API (i.e. rather than setting single values
+one-by-one, one should set an entire slice at once via a stacked tensor).
+These best practices will be shared in a blog post, as well as in the
+documentation for the new method(s).
+
+We may choose to deprecate existing set/get methods such as
+`tf.tensor_scatter_nd_update` in favor of the new API, since it is more general.
+In this case, a note can be made in the documentation for the deprecated methods
+with instructions on how to update.
+
+### Tutorials and Examples
+
+The update methods are meant as a drop-in replacement for numpy `__setitem__`.
+
+```python
+#======================================================================
+# Basic indexing
+#======================================================================
+
+# Single index.
+x = x.at[1, 2, 2].set(-1)
+
+# Slicing and striding.
+x = x.at[1, 1:4:2, 2].set(-1)
+
+#======================================================================
+# Advanced indexing
+#======================================================================
+
+# Integer array indexing.
+indices = [[0, 0, 1], [0, 4, 3], [3, 3, 2]]
+updates = [3, 4, 5]
+x = x.at[indices].set(updates)
+
+# Boolean array indexing.
+x = x.at[x < 0.5].set(0.5)
+```
+
+### Compatibility
+
+Since this change introduces a new API and does not add any ops, all backwards
+and forwards compatibility requirements will continue to be met.
+
+Similarly, all existing functionality should continue to work with other parts
+of the TensorFlow ecosystem (TFLite, GPU/TPU, SavedModel), as long as the ops
+themselves worked in the first place.
+
+### Third Parties
+
+This does not affect 3rd party partners, unless they wish to use the new API.
+
+### User Impact
+
+The proposed change will address the major pain point brought up by users
+(see [Motivation](#motivation)).  We will announce the feature via a blog post,
+and link to it on all the associated bugs.
+
+There are no special requirements for rolling out the feature.  The Python API
+can be added without changing existing models.
+
+## Detailed Design
+
+To introduce `.at`, we can use a Python `property` to allow it to treat
+`Tensor.at` as a function call.  This function returns a helper object that can
+be used with general indexing to update an underlying tensor:
+```python
+class TensorIndexUpdateHelper:
+  """Helper class that allows copying/modifying a tensor at a set of indices."""
+  def __init__(self, tensor, index_exp):
+    self.tensor = tensor
+    self.index_exp = index_exp
+
+  def set(self, values):
```

Comment:
Would it be helpful to add type annotations to these methods, e.g. `set(self, values) -> tf.Tensor`? This would duplicate some of your explanatory text below this code block, but that's not necessarily a bad thing if it improved clarity.

Reply:
Done.